synapse/http/replicationagent.py

```diff
@@ -0,0 +1,170 @@
+# Copyright 2023 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from typing import Optional
+
+from zope.interface import implementer
+
+from twisted.internet import defer
+from twisted.internet.endpoints import HostnameEndpoint, wrapClientTLS
+from twisted.internet.interfaces import IStreamClientEndpoint
+from twisted.python.failure import Failure
+from twisted.web.client import (
+    URI,
+    HTTPConnectionPool,
+    _AgentBase,
+    _DeprecatedToCurrentPolicyForHTTPS,
+)
+from twisted.web.error import SchemeNotSupported
+from twisted.web.http_headers import Headers
+from twisted.web.iweb import (
+    IAgent,
+    IAgentEndpointFactory,
+    IBodyProducer,
+    IPolicyForHTTPS,
+    IResponse,
+)
+
+from synapse.types import ISynapseReactor
+
+logger = logging.getLogger(__name__)
+
+
+@implementer(IAgentEndpointFactory)
+class ReplicationEndpointFactory:
+    """Connect to a given TCP socket"""
+
+    def __init__(
+        self,
+        reactor: ISynapseReactor,
+        context_factory: IPolicyForHTTPS,
+    ) -> None:
+        self.reactor = reactor
+        self.context_factory = context_factory
+
+    def endpointForURI(self, uri: URI) -> IStreamClientEndpoint:
+        """
+        This part of the factory decides what kind of endpoint is being connected to.
+
+        Args:
+            uri: The pre-parsed URI object containing all the uri data
+
+        Returns: The correct client endpoint object
+        """
+        if b"http" in uri.scheme:
+            endpoint = HostnameEndpoint(self.reactor, uri.host, uri.port)
+            if uri.scheme == b"https":
+                endpoint = wrapClientTLS(
+                    self.context_factory.creatorForNetloc(uri.host, uri.port), endpoint
+                )
+            return endpoint
```

Comment:
We'll need to mess with this logic in the future when adding unix support, but I think for now you want:

```suggestion
        endpoint = HostnameEndpoint(self.reactor, uri.host, uri.port)
        if uri.scheme == b"http":
            return endpoint
        elif uri.scheme == b"https":
            return wrapClientTLS(
                self.context_factory.creatorForNetloc(uri.host, uri.port), endpoint
            )
```

Alternately maybe the following would be less changes later:

```suggestion
        if uri.scheme in (b"http", b"https"):
            endpoint = HostnameEndpoint(self.reactor, uri.host, uri.port)
            if uri.scheme == b"https":
                endpoint = wrapClientTLS(
                    self.context_factory.creatorForNetloc(uri.host, uri.port), endpoint
                )
            return endpoint
```

Note the important thing is to remove the `b"http" in uri.scheme` as that will match things like `httpfoo` or other strings.

Reply:
I like option number 2 better, let's go with that.